api/v3/Payment.php

```diff
@@ -0,0 +1,291 @@
+}
+
+/**
+ * Delete a payment.
+ *
+ * @param array $params
+ *   Input parameters.
+ *
+ * @throws API_Exception
+ * @return array
+ *   Api result array
+ */
+function civicrm_api3_payment_delete(&$params) {
+  $params['id'] = isset($params['civicrm_financial_trxn_id']) ? $params['civicrm_financial_trxn_id'] : $params['payment_id'];
+  if (!CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_FinancialTrxn', $params['id'], 'id')) {
+    return civicrm_api3_create_error(ts('You need to supply a valid Payment ID to delete a payment'));
+  }
+  return civicrm_api3('FinancialTrxn', 'delete', $params);
+}
+
+/**
+ * Cancel/Refund a payment for a Contribution.
+ *
+ * @param array $params
+ *   Input parameters.
+ *
+ * @throws API_Exception
+ * @return array
+ *   Api result array
+ */
+function civicrm_api3_payment_cancel(&$params) {
+  $ftId = isset($params['civicrm_financial_trxn_id']) ? $params['civicrm_financial_trxn_id'] : $params['payment_id'];
+  if (!CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_FinancialTrxn', $ftId, 'id')) {
+    return civicrm_api3_create_error(ts('You need to supply a valid Payment ID to refund a payment'));
+  }
+  $eftParams = array(
+    'entity_table' => 'civicrm_contribution',
+    'financial_trxn_id' => $ftId,
+  );
+  $entity = civicrm_api3('EntityFinancialTrxn', 'get', $eftParams);
+  $entity = reset($entity['values']);
+  $contributionId = $entity['entity_id'];
+  $params['total_amount'] = $entity['amount'];
+  
+  $trxn = CRM_Contribute_BAO_Contribution::recordAdditionalPayment($contributionId, $params, 'refund', NULL, FALSE);
+  
+  $values = array(); 
+  if (is_a($trxn, 'CRM_Core_Error')) {
+    throw new API_Exception($trxn->_errors[0]['message']);
+  }
+  else {
+    _civicrm_api3_object_to_array_unique_fields($trxn, $values[$trxn->id]);
+  }
+  return civicrm_api3_create_success($values, $params, 'Payment', 'cancel', $trxn);
+}
+
+/**
+ * Add a payment for a Contribution.
+ *
+ * @param array $params
+ *   Input parameters.
+ *
+ * @throws API_Exception
+ * @return array
+ *   Api result array
+ */
+function civicrm_api3_payment_create(&$params) {
+  if (empty($params['contribution_id']) || 
+      (isset($params['contribution_id']) && !CRM_Core_DAO::getFieldValue('CRM_Contribute_DAO_Contribution', $params['contribution_id'], 'id'))) {
+    return civicrm_api3_create_error(ts('You need to supply a valid contribution ID to create a payment'));
+  }
+  // Check if it is an update
+  $ftId = isset($params['civicrm_financial_trxn_id']) ? $params['civicrm_financial_trxn_id'] : $params['payment_id'];
+  if ($ftId && (!CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_FinancialTrxn', $ftId, 'id') || !CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_FinancialTrxn', $ftId, 'is_payment'))) {
+    return civicrm_api3_create_error(ts('You need to supply a valid Payment ID to update a payment'));
+  }
+  elseif ($ftId) {
+    $amount = $params['total_amount'];
```

Comment:
Sure, the spec dictates that when a payment is cancelled, the original payment transaction is reversed and a new one is created. Here is the spec: http://wiki.civicrm.org/confluence/display/CRM/Payment+API